extra/migrate_settings.rb

```diff
@@ -1,108 +1,200 @@
+
+      migrations = @migrations.new_migrations
+      migrations.each do |migration|
+        m = migration.new(working_dir_path)
+
+        create_migration_dirs(m.src_dir, m.dst_dir)
+        if migration == migrations.first
+          copy_original_configuration(m.src_dir)
+        else
+          copy_previous_migration_results(@migrations.executed_migrations.last.dst_dir, m.src_dir)
+        end
+
+        m.migrate
+
+        if migration == migrations.last
+          copy_to_results_dir(m.dst_dir, result_dir_path)
+        end
+
+        @migrations.executed_migrations << m
+      end
+    ensure
+      @migrations.persist_migrations_state(result_dir_path)
+    end
+
+    def create_migration_dirs(src_dir, dst_dir)
+      FileUtils.mkdir_p(src_dir)
+      FileUtils.mkdir_p(File.join(dst_dir, "settings.d"))
+    end
+
+    def copy_original_configuration(src_dir)
+      FileUtils.cp_r(File.join(config_src_dir, '.'), src_dir)
+    end
+
+    def copy_previous_migration_results(result_dir, src_dir)
+      FileUtils.cp_r(File.join(result_dir, '.'), src_dir)
+    end
+
+    def copy_to_results_dir(dst_dir, result_dir)
+      FileUtils.cp_r(File.join(dst_dir, '.'), result_dir)
     end
   end
+end
 
-  return output, data
+def app_dir
+  path("../../", __FILE__)
 end
 
-def write_to_files(output,unknown)
-  modules.each do |m|
-    next if output[m] == {}
-    File.open("#{m}.yml",'w') do |f|
-      f.write(output[m].to_yaml)
-      if m == :settings && unknown != {}
-        f.write "\n# Unparsed options, please review\n"
-        f.write(unknown.to_yaml.gsub(/^---/,''))
-      end
+def path(part_one, part_two = nil)
+  File.expand_path(part_one, part_two)
+end
+
+def parse_cli_options(args)
+  result = {}
+
+  result[:src_dir] = path("config", app_dir)
+  result[:tmp_dir] = path("tmp", app_dir)
+  result[:migrations_dir] = path("migrations", path("extra", app_dir))
+  result[:migrations_state] = path("migration_state", path("config", app_dir))
+
+  parser = OptionParser.new do |opts|
+    opts.banner = "Usage: migrate.rb [options]"
+    opts.separator ""
+
+    opts.on('-s', '--src_dir SRC_DIR_PATH', 'configuration files directory path') do |src_dir|
+      result[:src_dir] = path(src_dir)
+    end
+    opts.on('-t', '--tmp_dir TMP_DIR_PATH', 'working directory path') do |tmp_dir|
+      result[:tmp_dir] = path(tmp_dir)
+    end
+    opts.on('-m', '--migrations_dir MIGRATIONS_DIR_PATH', 'migrations directory path') do |migrations_dir|
+      result[:migrations_dir] = path(migrations_dir)
+    end
+    opts.on('-ms', '--migration_state MIGRATION_STATE_PATH', 'path to the file storing executed migrations') do |ms_path|
+      result[:migrations_state] = path(ms_path)
     end
   end
+
+  parser.parse!(args)
+  result
 end
 
-# When running as a script
 if __FILE__ == $0 then
-  orig_file = ARGV[0] || '/etc/foreman-proxy/settings.yml'
-  data      = YAML.load_file(orig_file)
+  options = parse_cli_options(ARGV)
```

Comment:
Nit pick, this should error out when no options are provided, otherwise:

```
[lzap@lzapx smart-proxy-test]$ ruby extra/migrate_settings.rb
extra/migrate_settings.rb:17:in `load_past_migrations!': Migration state file /home/lzap/work/smart-proxy-test/config/migration_state can't be found (RuntimeError)
```